packages/common/solana/utils.ts

```diff
@@ -125,72 +142,125 @@ export async function signTransaction(invoker: Keypair | SignerWalletAdapter, tx
  * @param connection - Solana client connection
  * @param tx - Transaction instance
  * @param hash - blockhash information, the same hash should be used in the Transaction
+ * @param context - context at which blockhash has been retrieve
+ * @param commitment - optional commitment that will be used for simulation and confirmation
  * @returns Transaction signature
  */
 export async function executeTransaction(
   connection: Connection,
-  tx: Transaction,
-  hash: BlockhashWithExpiryBlockHeight,
+  tx: Transaction | VersionedTransaction,
+  { hash, context, commitment }: ConfirmationParams,
 ): Promise<string> {
-  const rawTx = tx.serialize();
+  if (!hash.lastValidBlockHeight || tx.signatures.length === 0 || !hash.blockhash) {
+    throw Error("Error with transaction parameters.");
+  }
+
+  for (let i = 0; i < 3; i++) {
+    let res: RpcResponseAndContext<SimulatedTransactionResponse>;
+    if (isTransactionVersioned(tx)) {
+      res = await connection.simulateTransaction(tx);
+    } else {
+      res = await connection.simulateTransaction(tx);
+    }
+    if (res.value.err) {
+      const errMessage = JSON.stringify(res.value.err);
+      if (!errMessage.includes("BlockhashNotFound") || i === 2) {
+        throw new SendTransactionError("failed to simulate transaction: " + errMessage, res.value.logs || undefined);
+      }
+    }
+    break;
+  }
 
-  if (!hash.lastValidBlockHeight || !tx.signature || !hash.blockhash) throw Error("Error with transaction parameters.");
+  const isVersioned = isTransactionVersioned(tx);
 
-  const signature = bs58.encode(tx.signature);
-  const confirmationStrategy: BlockheightBasedTransactionConfirmationStrategy = {
-    lastValidBlockHeight: hash.lastValidBlockHeight + 50,
-    signature,
-    blockhash: hash.blockhash,
-  };
-  try {
-    return await sendAndConfirmRawTransaction(connection, rawTx, confirmationStrategy);
-  } catch (e) {
-    // If BlockHeight expired, we will check tx status one last time to make sure
-    if (e instanceof TransactionExpiredBlockheightExceededError) {
-      await sleep(1000);
+  let signature: string;
+  if (isVersioned) {
+    signature = bs58.encode(tx.signatures[0]);
+  } else {
+    signature = bs58.encode(tx.signature!);
+  }
+
+  let blockheight = await connection.getBlockHeight(commitment);
+  let transactionSent = false;
+  const rawTransaction = tx.serialize();
+  while (blockheight < hash.lastValidBlockHeight) {
+    try {
+      await connection.sendRawTransaction(rawTransaction, {
+        maxRetries: 0,
+        minContextSlot: context.slot,
+        preflightCommitment: commitment,
+        skipPreflight: true,
+      });
+      transactionSent = true;
+    } catch (e) {
+      if (
+        transactionSent ||
+        (e instanceof SendTransactionError && e.message.includes("Minimum context slot has not been reached"))
+      ) {
+        continue;
+      }
+      throw e;
+    }
+    await sleep(500);
```

Comment:
why should it sleep every time ? May it skip sleep if `transactionSent === true`?